### PR TITLE
ZeRO1: Add bucketting logic to control the size of tensors for all-gather/reduce-scatter

### PR DESCRIPTION
**Pull request by jeffhataws**
This PR updates XLA ZeRO1 implementation to use [allgather coalesed](https://github.com/pytorch/xla/pull/5950) and [reduce-scatter coalesced](https://github.com/pytorch/xla/pull/5956).

### CODE REVIEW

**Comment by JackCaoG** [comment]
Test crashed at `	torch_xla::tensor_methods::all_gather_coalesced_out(std::vector<c10::intrusive_ptr<torch_xla::XLATensor, `  which seems to be a real issue.

File: torch_xla/core/xla_model.py [hunk 1]
```diff
@@ -614,20 +619,86 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
         raise TypeError(
             f"`output` needs to be a list of Tensors, but given {type(output)}."
         )
-      if len(output) != len(value):
+      if len(output) != len(input):
```
Comment (jeffhataws) [inline torch_xla/core/xla_model.py:622]: should be len(value)

File: torch_xla/core/xla_model.py [hunk 2]
```diff
@@ -614,20 +619,86 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
...
+        if total >= 0.5 * bucket_cap or (total + tensor_bytes) > 2 * bucket_cap:
+          out_tensors.extend(
+              _all_gather_coalesced(tensor_bucket, output_bucket))
+          out_tensors.extend(
+              _all_gather_coalesced([tensor],
+                                    [output_selected] if output else []))
+        else:
+          tensor_bucket.append(tensor)
+          if output != None:
+            output_bucket.append(output[i])
```
Comment (jeffhataws) [inline torch_xla/core/xla_model.py:676]: Should be [idx] not [i]

File: torch_xla/core/xla_model.py [hunk 3]
```diff
@@ -615,19 +620,100 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
...
+        output_selected = output[idx]
+        if tensor.numel() != output_selected.numel():
+          raise ValueError(
+              f"`output` tensor size doesn't match `input` tensor size for tensor list index {idx}: "
+              f"{output_selected.numel()} vs {tensor.numel()}.")
+
+      # Tensor is larger than bucket_cap, don't bucketize
+      if tensor_bytes > bucket_cap:
+        # Flush out previous buckets even if they don't fill up
+        if total >= 0.5 * bucket_cap or (total + tensor_bytes) > 2 * bucket_cap:
```
Comment (jeffhataws) [inline torch_xla/core/xla_model.py:676]: @aws-rhsoln in this if statement, if tensor_bytes is 500MB and bucket_cap is 160MB, event of total is 0 because buckets are empty, it would try to call _all_gather_coalesced on empty buckets, and crash?

File: torch_xla/core/xla_model.py [hunk 4]
```diff
@@ -615,19 +620,83 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
...
+      divisor = len(groups[0]) if type(groups[0]) == list else len(groups)
+    else:
+      divisor = xrt_world_size()
+    bucket_cap = bucket_cap / divisor
+
+    for idx, tensor in enumerate(value):
+      tensor_bytes = tensor.numel() * tensor.element_size()
+
+      # Tensor is larger than bucket_cap, don't bucketize
+      if tensor_bytes > bucket_cap:
```
Comment (jeffhataws) [inline torch_xla/core/xla_model.py:664]: Similar to https://github.com/pytorch/xla/pull/6417/files#r1511919328 ,  maybe this if section can be removed to simplify.

**Comment by JackCaoG** [comment]
@alanwaketan can you review this one as well since you also review the gradient bucketing one?

File: torch_xla/core/xla_model.py (hunk 3 above)
Comment (jeffhataws) [reply torch_xla/core/xla_model.py:666]: Also there are 2 magic numbers 0.5 and 2, making it difficult to test/reason. Again, we should remove.

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 5]
```diff
@@ -256,26 +270,52 @@ def step(self, closure=None, **kwargs):
...
+      for param_group, sharded_param_group in zip(
+          self.param_groups, self.base_optimizer.param_groups):
+        for param, shard in zip(param_group['params'],
+                                sharded_param_group['params']):
+          if param.grad is not None:
+            grad_shard = grad_shards[index]
+
+            if grad_shard.dtype != self.optimizer_dtype:
+              grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
           shard.grad = grad_shard
```
Comment (hgt312) [inline torch_xla/distributed/zero_redundancy_optimizer.py:317]: this 2 line not in `if param.grad is not None:`, need indents

Comment (jeffhataws) [reply torch_xla/distributed/zero_redundancy_optimizer.py]: fixed. thanks.

File: torch_xla/core/xla_model.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import io
 import itertools
 import logging
+import os
```
Comment (jeffhataws) [inline torch_xla/core/xla_model.py:4]: no longer needed. will remove.

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 7]
```diff
@@ -256,26 +274,53 @@ def step(self, closure=None, **kwargs):
...
+          bucket_cap_mb=self.bucket_cap_mb,
+      )
+      index = 0
+      for param_group, sharded_param_group in zip(
+          self.param_groups, self.base_optimizer.param_groups):
+        for param, shard in zip(param_group['params'],
+                                sharded_param_group['params']):
+          if param.grad is not None:
+            grad_shard = grad_shards[index]
+
```
Comment (hgt312) [inline torch_xla/distributed/zero_redundancy_optimizer.py:319]: delete empty line?

File: torch_xla/core/xla_model.py [hunk 8]
```diff
@@ -38,6 +38,9 @@
 
 XLA_LIB = Library("xla", "DEF")
 
+# Default bucket size for all-gather and reduce-scatter
+_ALL_GATHER_REDUCE_SCATTER_BUCKET_CAP_MB = 160
```
Comment (hgt312) [inline torch_xla/core/xla_model.py:42]: useless now?

**Review by hgt312 — APPROVED** [review]
overall LGTM

File: test/test_mp_all_gather.py [hunk 9]
```diff
@@ -95,6 +95,42 @@ def _mp_fn(index):
...
+            'xm.all_gather_bucketized() produced wrong reductions for item {i} in result list',
+            file=sys.stderr)
+        print(f'[{index}] {cpu_result}', file=sys.stderr)
+        sys.exit(1)
+
+    # Testing with a single replica group and tensor list as input and output!=None (out-of-place) (Bucketized)
+    # Reuse ordinal_tensors from previous test
+    output_tensors = [
+        torch.zeros([world_size], dtype=torch.float).to(device)
+        for i in range(input_list_size)
```
Comment (JackCaoG) [inline test/test_mp_all_gather.py:118]: nit `for _ in range` since you don't need `i`. or just do `[torch.zeros([world_size], dtype=torch.float).to(device)] * len(input_list_size)`

File: test/test_mp_all_gather.py [hunk 10]
```diff
@@ -95,6 +95,42 @@ def _mp_fn(index):
...
+    ]
+    # TODO: add support for list input with pin_layout=True and output!=None
+    result_list = xm.all_gather_bucketized(
+        ordinal_tensors, dim=0, output=output_tensors, pin_layout=False)
+
+    for i, result in enumerate(result_list):
```
Comment (JackCaoG) [inline test/test_mp_all_gather.py:124]: shouldn't you be checking the `output_tensors` instead of `result_list` here?

File: torch_xla/core/xla_model.py [hunk 11]
```diff
@@ -633,6 +636,97 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
...
+        if len(tensor_bucket):
+          out_tensors.extend(
+              _all_gather_coalesced(tensor_bucket, output_bucket))
+        total = 0
+        tensor_bucket = []
+        output_bucket = [] if output else None
+      total = total_new
+      tensor_bucket.append(tensor)
+      if output != None:
+        output_bucket.append(output[idx])
```
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:719]: this part can be refactored a bit it seems like you can refactor out
```
      tensor_bucket.append(tensor)
      if output != None:
        output_bucket.append(output[idx])
```
and
```
        total = 0
        tensor_bucket = []
        output_bucket = [] if output else None
```

it was a bit difficult for me to figure out the difference between these two branches when first reading it.. hopefully refactor would help.

File: torch_xla/core/xla_model.py [hunk 12]
```diff
@@ -846,6 +940,101 @@ def reduce_scatter(reduce_type,
...
+        if len(tensor_bucket):
+          out_tensors.extend(
+              _reduce_scatter_coalesced(tensor_bucket, output_bucket))
+        total = 0
+        tensor_bucket = []
+        output_bucket = [] if output else None
+      total = total_new
+      tensor_bucket.append(tensor)
+      if output != None:
+        output_bucket.append(output[idx])
```
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:1027]: ditto, consider refactoring

File: torch_xla/core/xla_model.py [hunk 13]
```diff
@@ -846,6 +940,101 @@ def reduce_scatter(reduce_type,
...
+  def _reduce_scatter_coalesced(_input_list, _output_list=None):
+    return reduce_scatter(
+        reduce_type=reduce_type,
+        input=_input_list,
+        scale=scale,
+        scatter_dim=scatter_dim,
+        shard_count=shard_count,
+        groups=groups,
+        output=_output_list,
+        pin_layout=pin_layout)
```
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:992]: is this the only difference between `reduce_scatter_bucketized ` and `all_gather_bucketized`, if it is we can even further refactor the implementation of these two methods.

**Review by JackCaoG — COMMENTED** [review]
mostly lgtm, minor comments.

File: test/test_mp_all_gather.py (hunk 10 above)
Comment (jeffhataws) [reply test/test_mp_all_gather.py:124]: I believe output is also returned so result_list should be = output_tensors.

https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L599

File: torch_xla/core/xla_model.py (hunk 8 above)
Comment (jeffhataws) [reply torch_xla/core/xla_model.py:42]: yes

File: torch_xla/core/xla_model.py (hunk 11 above)
Comment (jeffhataws) [reply torch_xla/core/xla_model.py]: refactored

File: torch_xla/core/xla_model.py (hunk 13 above)
Comment (jeffhataws) [reply torch_xla/core/xla_model.py:997]: refactored

File: torch_xla/core/xla_model.py (hunk 12 above)
Comment (jeffhataws) [reply torch_xla/core/xla_model.py]: refactored

File: test/test_mp_all_gather.py (hunk 9 above)
Comment (jeffhataws) [reply test/test_mp_all_gather.py:118]: Something went wrong when I tried the _ change. So I kept this as is.

**Comment by JackCaoG** [comment]
@jeffhataws is this ready for another round of review?

**Comment by jeffhataws** [comment]
> @jeffhataws is this ready for another round of review?

I have a set of cleanup coming in an hour or so. I noticed that we have this code which was unique in all-gather. I will remove it, and have separate bucket_cap_mb for allgather and reduce-scatter in ZeRO1.
```
    if groups:
      divisor = len(groups[0]) if type(groups[0]) == list else len(groups)
    else:
      divisor = xrt_world_size()
    self._bucket_cap = self._bucket_cap / divisor
```

**Comment by jeffhataws** [comment]
> @jeffhataws is this ready for another round of review?

It is ready now for another round. Thanks.

**Review by JackCaoG — APPROVED** [review]

**Comment by JackCaoG** [comment]
@jeffhataws Thanks for the refactoring work!

File: test/test_mp_reduce_scatter.py [hunk 14]
```diff
@@ -83,6 +110,66 @@ def _mp_fn(index):
...
+      expected_world = xm.all_reduce(xm.REDUCE_SUM, xrand_list[i], scale)
+      xm.mark_step()
+
+      slice_idx = torch.tensor(
+          list(range(index * shard_size, (index + 1) * shard_size)))
+      expected = expected_world.cpu().index_select(scatter_dim, slice_idx)
+      assert res.cpu().allclose(expected)
+
+    xm.rendezvous(
+        'test_reduce_scatter_list_input_output_bucketized, zero bucket size')
```
Comment (jeffhataws) [inline test/test_mp_reduce_scatter.py:172]: Hi @JackCaoG , does rendezvous allow comma and space in the rendezvous key? How come this didn't error out?

Comment (jeffhataws) [reply test/test_mp_reduce_scatter.py:172]: If this is not a concern, we can merge this PR.

Comment (JackCaoG) [reply test/test_mp_reduce_scatter.py:172]: Looking at the implementation of `xla_rendezvous`, I think tag got ignored so it doesn't really matter.
https://github.com/pytorch/xla/blob/782f05d2f2656065663d432b5df9e149409719a3/torch_xla/core/xla_model.py#L1110-L1115
